### PR TITLE
String funcs

### PR DESCRIPTION
**Pull request by AviAvni**
implement replace function https://github.com/RedisGraph/RedisGraph/issues/1965

### CODE REVIEW

File: src/arithmetic/string_funcs/string_funcs.c [hunk 1]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+	const char *old_string = argv[1].stringval;
+	const char *new_string = argv[2].stringval;
+	size_t old_string_len = strlen(old_string);
+	size_t new_string_len = strlen(new_string);
+
+	if(old_string_len == 0) {
+		ErrorCtx_RaiseRuntimeException("ArgumentError: search string can't be length 0");
+		// Incase expection handler wasn't set, return NULL.
+		return SI_NullVal();
+	}
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:259]: is this case covered by a test ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 2]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+	char *ptr = str;
+
+	char **arr = array_new(char *, 0);
+
+	while(true) {
+        // find pointer to next substring
+        ptr = strstr(ptr, old_string);
+
+        // if no substring found, then break from the loop
+        if(ptr == NULL) { break; }
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:270]: Why the curly braces ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 3]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+
+		// store ptr for replace use
+		array_append(arr, ptr);
+
+        // increment our string pointer
+        ptr += old_string_len;
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:276]: indentation

File: src/arithmetic/string_funcs/string_funcs.c [hunk 4]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+		return SI_NullVal();
+	}
+
+	char *ptr = str;
+
+	char **arr = array_new(char *, 0);
+
+	while(true) {
+        // find pointer to next substring
+        ptr = strstr(ptr, old_string);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:267]: just making sure, if `ptr` is pointing to `NULL` then `NULL` is returned ?

File: src/arithmetic/string_funcs/string_funcs.c [hunk 5]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+
+		// store ptr for replace use
+		array_append(arr, ptr);
+
+        // increment our string pointer
+        ptr += old_string_len;
+    }
+
+	int occurances = array_len(arr);
+	if(occurances == 0) return argv[0];
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:280]: mustn't do that, it is assumed we're returning a new string which will be freed at a later point in time

File: src/arithmetic/string_funcs/string_funcs.c [hunk 6]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+
+	// calculate new buffer size
+	size_t buffer_size = strlen(str) + occurances * new_string_len - occurances * old_string_len;
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:283]: `strlen(str) + (occurrences * new_string_len) - (occurrences * old_string_len);`

File: src/arithmetic/string_funcs/string_funcs.c [hunk 7]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+        if(ptr == NULL) { break; }
+
+		// store ptr for replace use
+		array_append(arr, ptr);
+
+        // increment our string pointer
+        ptr += old_string_len;
+    }
+
+	int occurances = array_len(arr);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:279]: `occurrences`

File: src/arithmetic/string_funcs/string_funcs.c [hunk 8]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+	if(occurances == 0) return argv[0];
+
+	// calculate new buffer size
+	size_t buffer_size = strlen(str) + occurances * new_string_len - occurances * old_string_len;
+
+	// allocate buffer
+	char *buffer = (char*) malloc(buffer_size*sizeof(char) + 1);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:286]: Switch from the malloc family to `rm_malloc`

File: src/arithmetic/string_funcs/string_funcs.c [hunk 9]
```diff
@@ -241,6 +241,83 @@ SIValue AR_ENDSWITH(SIValue *argv, int argc) {
...
+		// move forwart to copy more data from the original string
+		ptr = arr[i] + old_string_len;
+	}
+
+	// copy rest of the string from the original string
+	strcpy(buffer_ptr, ptr);
+
+	buffer[buffer_size] = '\0';
+
+	return SI_ConstStringVal(buffer);
```
Comment (swilly22) [inline src/arithmetic/string_funcs/string_funcs.c:318]: `return SI_StringVal(buffer);`
the string needs to be free at some point.

File: tests/unit/test_arithmetic_expression.cpp [hunk 10]
```diff
@@ -1719,3 +1719,21 @@ TEST_F(ArithmeticTest, CoalesceTest) {
 	ASSERT_EQ(0, SIValue_Compare(SI_LongVal(1), arExp->operand.constant, NULL));
 }
 
+TEST_F(ArithmeticTest, ReplaceTest) {
+	const char *query;
+	AR_ExpNode *arExp;
+
+	// Test replace in start of the string.
```
Comment (swilly22) [inline tests/unit/test_arithmetic_expression.cpp:1726]: missing a case where the old string doesn't exists in the original string
please add a test case where new string is empty ''

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]
